File: client/src/types/dataExtract.ts
export type DataExtractFormat = "csv" | "json" | "xml" | "delimited";

export type XmlExtractMode = "rows" | "tree";

export type DelimitedProfile = "pipe_embed" | "custom";

export interface ExtractionResult {
  format: DataExtractFormat;
  records: Record<string, unknown>[];
  meta: Record<string, unknown>;
}

export interface ImportRowError {
  line_number?: number | null;
  message: string;
  raw_line?: string | null;
}

export interface ImportResult {
  batch_id: number;
  format: string;
  profile?: string | null;
  row_count: number;
  error_count: number;
  preview: Record<string, unknown>[];
  errors: ImportRowError[];
}

export interface ImportBatchSummary {
  id: number;
  filename: string;
  format: string;
  profile?: string | null;
  status: "completed" | "partial" | "failed";
  row_count: number;
  error_count: number;
  imported_by: number;
  promoted_count: number;
  promoted_at?: string | null;
  meta: Record<string, unknown>;
}

export interface ImportBatchList {
  items: ImportBatchSummary[];
  total: number;
}

export interface ImportBatchDetail {
  batch: ImportBatchSummary;
  preview_rows: Array<{
    id: number;
    batch_id: number;
    row_index: number;
    fields: Record<string, unknown>;
    raw_line?: string | null;
    error?: string | null;
  }>;
}

export interface PromoteBatchResult {
  batch_id: number;
  promoted: number;
  skipped: number;
  errors: string[];
}

File: server/app/schemas/data_extract.py
"""Schemas for data extraction and import API."""

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.common import PaginatedResponse


class ExtractionResultResponse(BaseModel):
    format: Literal["csv", "json", "xml", "delimited"]
    records: list[Any]
    meta: dict[str, Any] = Field(default_factory=dict)


class ImportRowErrorResponse(BaseModel):
    line_number: int | None = None
    message: str
    raw_line: str | None = None


class ImportResultResponse(BaseModel):
    batch_id: int
    format: str
    profile: str | None = None
    row_count: int
    error_count: int
    preview: list[dict[str, Any]] = Field(default_factory=list)
    errors: list[ImportRowErrorResponse] = Field(default_factory=list)


class ImportBatchResponse(BaseModel):
    id: int
    filename: str
    format: str
    profile: str | None = None
    status: Literal["completed", "partial", "failed"]
    row_count: int
    error_count: int
    imported_by: int
    promoted_count: int = 0
    promoted_at: datetime | None = None
    meta: dict[str, Any] = Field(default_factory=dict)

    model_config = ConfigDict(from_attributes=True)


class ImportBatchListResponse(PaginatedResponse[ImportBatchResponse]):
    """Paginated import batch list."""


class PromoteBatchResponse(BaseModel):
    batch_id: int
    promoted: int
    skipped: int
    errors: list[str] = Field(default_factory=list)


class ImportRowResponse(BaseModel):
    id: int
    batch_id: int
    row_index: int
    fields: dict[str, Any] = Field(default_factory=dict)
    raw_line: str | None = None
    error: str | None = None

    model_config = ConfigDict(from_attributes=True)


class ImportBatchDetailResponse(BaseModel):
    batch: ImportBatchResponse
    preview_rows: list[ImportRowResponse] = Field(default_factory=list)
